src/redprl/refiner_composition_kit.fun

```diff
@@ -188,16 +191,114 @@ struct
         val (I, H) >> CJ.EQ ((hcom, other), ty) = jdg
         val Syn.HCOM {dir=(r, r'), ty=ty0, cap, tubes} = Syn.out hcom
         val (eq, (u, tube)) = Option.valOf (List.find (fn (eq, _) => P.eq Sym.eq eq) tubes)
-        val () = assertAlphaEq (ty0, ty)
-  
-        val (goalTy, _) = makeGoal @@ (I, H) >> CJ.TYPE ty
-        val (goalCap, _) = makeGoal @@ (I, H) >> CJ.MEM (cap, ty)
-        val (goalEq, _) = makeGoal @@ (I, H) >> CJ.EQ ((substSymbol (r', u) tube, other), ty)
+
+        (* type *)
+        val goalTy0 = makeGoal' @@ (I, H) >> CJ.EQ_TYPE (ty0, ty)
+
+        (* cap *)
+        val goalCap = makeGoal' @@ (I, H) >> CJ.MEM (cap, ty)
+
+        (* eq *)
+        val goalEq = makeGoal' @@ (I, H) >> CJ.EQ ((substSymbol (r', u) tube, other), ty)
   
         val w = alpha 0
       in
         T.empty
-          >: goalTy >: goalCap >: goalEq
+          >: goalTy0 >: goalCap >: goalEq
+          >:+ ComKit.genInterTubeGoals (I, H) w ty tubes tubes
+          >:+ ComKit.genTubeCapGoals (I, H) ty r cap tubes
+        #> (I, H, trivial)
+      end
+
+    val TubeEqR = catJdgFlipWrapper TubeEqL
+
+    (* Try all the hcom rules.
+     * Note that the EQ rule is invertible only when the cap and tube rules fail. *)
+    val AutoEqLR = CapEqL orelse_ CapEqR orelse_ TubeEqL orelse_ TubeEqR orelse_ Eq
+    val AutoEqL = CapEqL orelse_ TubeEqL orelse_ Eq
+    val AutoEqR = CapEqR orelse_ TubeEqR orelse_ Eq
+  end
+
+  structure Com =
```

Comment:
@favonia I agree that we can fix them later if they are not quite right... Once we add tests, it may become clear whether these are right or wrong, so we can wait until then.

Ultimately, I expect that @cangiuli would write something down about this at some point, so we can just compare the rules that he writes down with what is implemented, when it is time.